File: homeworks/Homework2/bisection.cpp
#include<iostream>
#include<cmath>
#include<cstdlib>


double bisection(double (*f)(x), double a, double b, double tol){
	double root, zero, c, f_a, f_b, f_c;

	f_a = f(a);
	f_b = f(b);

	zero = 0.0;

	//base case checking
	if (f_a*f_b > zero){
		std::cout << "Unable to determine a root from the endpoints given. " << std::endl << "Please pick two values on f(x) that reside on opposite sides of the x-axis." << std::endl << std::endl << "Thank you. Now exiting...";
		return EXIT_FAILURE;
	}

	else if (f_a*f_b == zero){
		if (f_a == zero) 
			return a;
		else 
			return b;
	}

	if (tol <= 0){
		std::cout << "Unable to process a tolerance of " << tol << std::endl << "Please make sure it is > 0" << std::endl << "Now exiting...";

		return EXIT_FAILURE;
	}




	double k = ceil((log2(tol / abs(b - a))) / log2(0.5) + 1);

	for(int i = 0; i<k; i++){
		c = 0.5*(b+a);
		f_c = f(c);

		if ((f_a*f_c) < 0){
			b = c;
			f_b = f_c;
		}
		else{
			a = c;
			f_a = f_c;
		}
	}
	
	root = c;

	return root;
}






double func_a(double x){
	return (pow(x,2) - 3);
}

double func_b(double x){
	long double PI = 3.14159265358979323846264338327950288419716939937510;
	return (sin(PI*x));
}


int main(){

	std::cout << "a) " << bisection(&func_a, 1, 4, 0.005) << std::endl;
	std::cout << "b) " << bisection(&func_b, 0.001, 3, 0.005) << std::endl;
	

	return 0;
}

File: homeworks/Homework2/bisection.py
import math
import sys


def func_a(x):
    return (pow(x,2) -3)

def func_b(x):
    PI = 3.14159265358979323846264338327950288419716939937510
    return math.sin(PI*x)

def bisection(function,a,b,tol):
    f_a = function(a)
    f_b = function(b)
    zero = 0.0

    if tol <= 0:
        print(f"Unable to process a tolerance of {tol}")
        print("Please make sure tolerance > 0")
        sys.exit(1)

    if f_a*f_b > zero:
        print("Unable to determine a root from the given endpoints")
        print("Please pick two values on f(x) that reside on opposite sides of the x-axis")
        sys.exit(1)

    elif f_a*f_b == zero:
        if f_a == zero:
            return a
        else:
            return b

    k = math.ceil((math.log(tol/math.fabs(b-a),2)) / math.log(0.5,2) + 1)

    for i in range(k):
        c = 0.5 * (b+a)
        f_c = function(c)

        if(f_a*f_c) < 0:
            b = c
            f_b = f_c
        else:
            a = c
            f_a = f_c

    return c


print(f"a) {bisection(func_a,1,4,0.005)}")
print(f"b) {bisection(func_b,0.001,1.5,0.005)}")